module/os/freebsd/zfs/zvol_os.c

```diff
@@ -1490,11 +1491,12 @@ zvol_os_create_minor(const char *name)
 	zv->zv_zilog = zil_open(os, zvol_get_data, &zv->zv_kstat.dk_zil_sums);
 	if (spa_writeable(dmu_objset_spa(os))) {
 		if (zil_replay_disable)
-			zil_destroy(zv->zv_zilog, B_FALSE);
+			replayed_zil = zil_destroy(zv->zv_zilog, B_FALSE);
 		else
-			zil_replay(os, zv, zvol_replay_vector);
+			replayed_zil = zil_replay(os, zv, zvol_replay_vector);
 	}
-	zil_close(zv->zv_zilog);
+	if (replayed_zil)
+		zil_close(zv->zv_zilog);
 	zv->zv_zilog = NULL;
```

Comment:
So you are not closing zil but clearing zv->zv_zilog so that it was open again?  zil_open() seems to be quite invariant now, but wouldn't it be better to leave zv->zv_zilog as-is?  I may be wrong, haven't looked too deep.

Reply:
`zil_open` doesn't actually have any side effects.  All it does is return a pointer to that zilog_t field, and set its callback memebers.  So simply nulling it out doesn't leak anything.  I didn't consider the possibility of not nulling it out, because the existing code always nulls it.

Reply:
I agree that second open is not a problem now, but still not good.  Also actual zil_close() is not so small, it it even to be called, and I am not sure how much invariant it is.

Reply:
I've found what looks like identical logic in zvol_shutdown_zv().  So while I still see it as ugly, I have no objections.